measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
+        # compute FIM as Np*Np
+        fim = np.zeros((self.no_param, self.no_param))
+
+        # go over all measurement index
+        for m1_rank in range(self.total_no_measure):
+            # get the corresponding gradient vector for this measurement
+            # use np.matrix to keep the dimensions for the vector, otherwise numpy makes the vector 1D instead of Np*1
+            jac_m1 = np.matrix(self.jac[m1_rank])
+            # go over all measurement index
+            for m2_rank in range(self.total_no_measure):
+                # get the corresponding gradient vector for this measurement
+                jac_m2 = np.matrix(self.jac[m2_rank])
+                # solution of if these two measurements are selected
+                measure_matrix = np.matrix([measurement_matrix[m1_rank, m2_rank]])
+                # retrieve the error covariance matrix corresponding part
+                sigma = np.matrix([self.Sigma_inv[m1_rank, m2_rank]])
+                # compute FIM as Q.T@y@error_cov@Q
+                fim_unit = jac_m1.T @ measure_matrix @ sigma @ jac_m2
+                fim += fim_unit
         # FIM read
         if self.verbose:
-            self.__print_FIM(FIM)
+            print_fim(fim)
 
-        return FIM
+        return fim
 
-    def __measure_matrix(self, measurement_vector):
+    def solve(
+        self, mip_option=False, objective=ObjectiveLib.A, degeneracy_hunter=False
+    ):
         """
+        Set up solvers, solve the problem, and check the solver status and termination conditions
 
-        :param measurement_vector:
-        :return:
+        Arguments
+        ---------
+        :mip_option: boolean, if True, it is a mixed-integer problem, otherwise it is a relaxed problem with no integer decisions
+        :objective: Enum, "A" or "D" optimality, use trace or determinant of FIM
+        :degeneracy_hunter: boolean, when set up to True, use degeneracy hunter to check infeasibility in constraints. For debugging.
+
+        Return
+        ------
+        None
         """
-        # check if measurement vector legal
-        assert len(measurement_vector)==self.total_no_measure, "Measurement vector is of wrong shape!!!"
+        if self.fixed_nlp:
+            solver = pyo.SolverFactory("cyipopt")
+            solver.config.options["hessian_approximation"] = "limited-memory"
+            additional_options = {
+                "max_iter": 3000,
+                "output_file": "console_output",
+                "linear_solver": "mumps",
+                # "halt_on_ampl_error": "yes", # this option seems not working for cyipopt
+                "bound_push": 1e-10,
+            }
 
-        measurement_matrix = np.zeros((self.total_no_measure, self.total_no_measure))
+            if degeneracy_hunter:
+                additional_options = {
+                    "max_iter": 0,
+                    "output_file": "console_output",
+                    "linear_solver": "mumps",
```

Comment:
Why are you specifying the linear solver? Why not just rely on the default linear solver? If someone has HSL installed, ma27 is the default. Otherwise, mumps is the default.

Reply:
added toggle